In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os
path = r'/Users/jakubkondelka/Documents/CF_Instacart Basket Analysis'


In [4]:
op_merged = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_48.pkl'))

In [8]:
# check w head 
op_merged.head()

order_id  user_id evaluations  order_number  orders_day_of_week  \
0   2539329        1       prior             1                   2   
1   2398795        1       prior             2                   3   
2    473747        1       prior             3                   3   
3   2254736        1       prior             4                   4   
4    431534        1       prior             5                   4   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     NaN         196                  1   
1                  7                    15.0         196                  1   
2                 12                    21.0         196                  1   
3                  7                    29.0         196                  1   
4                 15                    28.0         196                  1   

   reordered  ... aisle_id  department_id  prices  _merge    price_range_loc  \
0          0  ...       77              7     9.0    both  Mid-range product   
1          1  ...       77              7     9.0    both  Mid-range product   
2          1  ...       77              7     9.0    both  Mid-range product   
3          1  ...       77              7     9.0    both  Mid-range product   
4          1  ...       77              7     9.0    both  Mid-range product   

      busiest day    Busiest days Busiest_period_of_day max_order  \
0  Regularly busy  Regularly busy           Most orders        10   
1  Regularly busy      Least busy         Fewest orders        10   
2  Regularly busy      Least busy           Most orders        10   
3      Least busy      Least busy         Fewest orders        10   
4      Least busy      Least busy           Most orders        10   

   loyalty_flag  
0  New customer  
1  New customer  
2  New customer  
3  New customer  
4  New customer  

[5 rows x 21 columns]

In [7]:
# check shape
op_merged.shape

(32404859, 21)

## 2) Find the aggregated mean of the “order_number” column grouped by “department_id”

In [9]:
# aggregate by MEAN of order_number and group by column department_id
op_merged.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

The results of running this on full data frame vs a subset is that hear every department is grouped by and each has a mean of order numbers. The subset had a size of only milion rows, therefore not all departments were part of the result.
We can also see that maximum MEAN of subset was 18,5 and with full data set it is 22.9

## 5) Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). 

In [16]:
# I want to group by Loyalty category and find min,max,mean of product prices 
op_merged.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

we can see the loayl customers buy on average the cheapest products. Than it is regular customer and the new customer is spending the most

## 6) The team now wants to target different types of spenders in their marketing campaigns. This can be achieved by looking at the prices of the items people are buying. Create a spending flag for each user based on the average price across all their orders using the following criteria:
If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.”
If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”

In [22]:
# use transform to aggregate prices by Mean. Group by user_id and prices

op_merged['mean_prices'] = op_merged.groupby(['user_id'])['prices'].transform(np.mean)

In [28]:
op_merged.head()

order_id  user_id evaluations  order_number  orders_day_of_week  \
0   2539329        1       prior             1                   2   
1   2398795        1       prior             2                   3   
2    473747        1       prior             3                   3   
3   2254736        1       prior             4                   4   
4    431534        1       prior             5                   4   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     NaN         196                  1   
1                  7                    15.0         196                  1   
2                 12                    21.0         196                  1   
3                  7                    29.0         196                  1   
4                 15                    28.0         196                  1   

   reordered  ... prices  _merge    price_range_loc     busiest day  \
0          0  ...    9.0    both  Mid-range product  Regularly busy   
1          1  ...    9.0    both  Mid-range product  Regularly busy   
2          1  ...    9.0    both  Mid-range product  Regularly busy   
3          1  ...    9.0    both  Mid-range product      Least busy   
4          1  ...    9.0    both  Mid-range product      Least busy   

     Busiest days Busiest_period_of_day max_order  loyalty_flag max_prices  \
0  Regularly busy           Most orders        10  New customer   6.367797   
1      Least busy         Fewest orders        10  New customer   6.367797   
2      Least busy           Most orders        10  New customer   6.367797   
3      Least busy         Fewest orders        10  New customer   6.367797   
4      Least busy           Most orders        10  New customer   6.367797   

   mean_prices  
0     6.367797  
1     6.367797  
2     6.367797  
3     6.367797  
4     6.367797  

[5 rows x 23 columns]

In [31]:
# if condtions to flag based on High Spender - mean > 10. Low Spender - mean < 10

op_merged.loc[op_merged['mean_prices'] < 10, 'Spender_flag'] = 'Low Spender'
op_merged.loc[op_merged['mean_prices'] > 10, 'Spender_flag'] = 'High Spender'

In [34]:
#check head
op_merged.head(100)

order_id  user_id evaluations  order_number  orders_day_of_week  \
0    2539329        1       prior             1                   2   
1    2398795        1       prior             2                   3   
2     473747        1       prior             3                   3   
3    2254736        1       prior             4                   4   
4     431534        1       prior             5                   4   
..       ...      ...         ...           ...                 ...   
95   3226575      360       prior             1                   5   
96   1469869      377       prior             3                   5   
97   1927023      387       prior             2                   4   
98    858092      420       prior             4                   1   
99    119314      420       prior             5                   3   

    order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                   8                     NaN         196                  1   
1                   7                    15.0         196                  1   
2                  12                    21.0         196                  1   
3                   7                    29.0         196                  1   
4                  15                    28.0         196                  1   
..                ...                     ...         ...                ...   
95                 12                     NaN         196                  1   
96                 17                     3.0         196                  9   
97                 10                    22.0         196                  3   
98                 19                    30.0         196                  2   
99                 19                    16.0         196                  2   

    reordered  ... _merge    price_range_loc     busiest day    Busiest days  \
0           0  ...   both  Mid-range product  Regularly busy  Regularly busy   
1           1  ...   both  Mid-range product  Regularly busy      Least busy   
2           1  ...   both  Mid-range product  Regularly busy      Least busy   
3           1  ...   both  Mid-range product      Least busy      Least busy   
4           1  ...   both  Mid-range product      Least busy      Least busy   
..        ...  ...    ...                ...             ...             ...   
95          0  ...   both  Mid-range product  Regularly busy  Regularly busy   
96          0  ...   both  Mid-range product  Regularly busy  Regularly busy   
97          0  ...   both  Mid-range product      Least busy      Least busy   
98          0  ...   both  Mid-range product  Regularly busy     Busiest day   
99          1  ...   both  Mid-range product  Regularly busy      Least busy   

   Busiest_period_of_day max_order      loyalty_flag max_prices mean_prices  \
0            Most orders        10      New customer   6.367797    6.367797   
1          Fewest orders        10      New customer   6.367797    6.367797   
2            Most orders        10      New customer   6.367797    6.367797   
3          Fewest orders        10      New customer   6.367797    6.367797   
4            Most orders        10      New customer   6.367797    6.367797   
..                   ...       ...               ...        ...         ...   
95           Most orders         3      New customer  10.006250   10.006250   
96           Most orders         3      New customer   8.496552    8.496552   
97           Most orders         8      New customer   7.396610    7.396610   
98        Average orders        22  Regular customer   7.387805    7.387805   
99        Average orders        22  Regular customer   7.387805    7.387805   

    Spender_flag  
0    Low Spender  
1    Low Spender  
2    Low Spender  
3    Low Spender  
4    Low Spender  
..           ...  
95  High Spender  
96   Low Spender  
97   Low Spender  
98   Low Spender  
99   Low Spender  

[100 rows x 24 columns]

In [35]:
#check shape
op_merged.shape

(32404859, 24)

In [36]:
# delete max_prices columns - it was a mistake 
op_merged = op_merged.drop(columns = ['max_prices'])

In [38]:
# check head
op_merged.head()

order_id  user_id evaluations  order_number  orders_day_of_week  \
0   2539329        1       prior             1                   2   
1   2398795        1       prior             2                   3   
2    473747        1       prior             3                   3   
3   2254736        1       prior             4                   4   
4    431534        1       prior             5                   4   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     NaN         196                  1   
1                  7                    15.0         196                  1   
2                 12                    21.0         196                  1   
3                  7                    29.0         196                  1   
4                 15                    28.0         196                  1   

   reordered  ... prices  _merge    price_range_loc     busiest day  \
0          0  ...    9.0    both  Mid-range product  Regularly busy   
1          1  ...    9.0    both  Mid-range product  Regularly busy   
2          1  ...    9.0    both  Mid-range product  Regularly busy   
3          1  ...    9.0    both  Mid-range product      Least busy   
4          1  ...    9.0    both  Mid-range product      Least busy   

     Busiest days Busiest_period_of_day max_order  loyalty_flag mean_prices  \
0  Regularly busy           Most orders        10  New customer    6.367797   
1      Least busy         Fewest orders        10  New customer    6.367797   
2      Least busy           Most orders        10  New customer    6.367797   
3      Least busy         Fewest orders        10  New customer    6.367797   
4      Least busy           Most orders        10  New customer    6.367797   

   Spender_flag  
0   Low Spender  
1   Low Spender  
2   Low Spender  
3   Low Spender  
4   Low Spender  

[5 rows x 23 columns]

In [51]:
# value counts
op_merged['Spender_flag'].value_counts(dropna = False)

Low Spender     31770614
High Spender      633864
NaN                  381
Name: Spender_flag, dtype: int64

## 7) In order to send relevant notifications to users within the app (for instance, asking users if they want to buy the same item again), the Instacart team wants you to determine frequent versus non-frequent customers. Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. The criteria for the flag should be as follows:
If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer.”
If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer.”
If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”

In [40]:
# Create an order frequency flag that marks the regularity of a user’s ordering behavior according
# to the median in the “days_since_prior_order” column.



In [41]:
op_merged['order_frequency'] = op_merged.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [43]:
# head
op_merged.head(20)

order_id  user_id evaluations  order_number  orders_day_of_week  \
0    2539329        1       prior             1                   2   
1    2398795        1       prior             2                   3   
2     473747        1       prior             3                   3   
3    2254736        1       prior             4                   4   
4     431534        1       prior             5                   4   
5    3367565        1       prior             6                   2   
6     550135        1       prior             7                   1   
7    3108588        1       prior             8                   1   
8    2295261        1       prior             9                   1   
9    2550362        1       prior            10                   4   
10   2968173       15       prior            15                   1   
11   1870022       15       prior            17                   2   
12   1911383       15       prior            18                   2   
13   2715276       15       prior            21                   1   
14    487368       15       prior            22                   1   
15   2293453       19       prior             2                   5   
16   1973799       19       prior             5                   6   
17    532817       19       prior             7                   4   
18   1573906       21       prior            10                   3   
19   1593000       31       prior            10                   3   

    order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                   8                     NaN         196                  1   
1                   7                    15.0         196                  1   
2                  12                    21.0         196                  1   
3                   7                    29.0         196                  1   
4                  15                    28.0         196                  1   
5                   7                    19.0         196                  1   
6                   9                    20.0         196                  1   
7                  14                    14.0         196                  2   
8                  16                     0.0         196                  4   
9                   8                    30.0         196                  1   
10                  9                     7.0         196                  2   
11                 16                     8.0         196                  6   
12                 11                     7.0         196                  1   
13                  9                     7.0         196                  1   
14                 10                    14.0         196                  1   
15                 14                     6.0         196                  3   
16                 12                     8.0         196                 15   
17                 17                     6.0         196                  1   
18                 10                     6.0         196                  2   
19                  8                     7.0         196                 17   

    reordered  ... _merge    price_range_loc     busiest day    Busiest days  \
0           0  ...   both  Mid-range product  Regularly busy  Regularly busy   
1           1  ...   both  Mid-range product  Regularly busy      Least busy   
2           1  ...   both  Mid-range product  Regularly busy      Least busy   
3           1  ...   both  Mid-range product      Least busy      Least busy   
4           1  ...   both  Mid-range product      Least busy      Least busy   
5           1  ...   both  Mid-range product  Regularly busy  Regularly busy   
6           1  ...   both  Mid-range product  Regularly busy     Busiest day   
7           1  ...   both  Mid-range product  Regularly busy     Busiest day   
8           1  ...   both  Mid-range product  Regularly busy     Busiest day   
9           1  ...   both  Mid-

In [46]:
# if condtions to flag based regularity of visits

op_merged.loc[op_merged['order_frequency'] > 20, 'frequency_flag'] = 'Non frequent customer'
op_merged.loc[(op_merged['order_frequency'] <= 20) & (op_merged['order_frequency'] > 10), 'frequency_flag'] = 'Regular customer'
op_merged.loc[op_merged['order_frequency'] <= 10, 'frequency_flag'] = 'Frequent customer'

In [47]:
# chekc head 
op_merged.head(100)

order_id  user_id evaluations  order_number  orders_day_of_week  \
0    2539329        1       prior             1                   2   
1    2398795        1       prior             2                   3   
2     473747        1       prior             3                   3   
3    2254736        1       prior             4                   4   
4     431534        1       prior             5                   4   
..       ...      ...         ...           ...                 ...   
95   3226575      360       prior             1                   5   
96   1469869      377       prior             3                   5   
97   1927023      387       prior             2                   4   
98    858092      420       prior             4                   1   
99    119314      420       prior             5                   3   

    order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                   8                     NaN         196                  1   
1                   7                    15.0         196                  1   
2                  12                    21.0         196                  1   
3                   7                    29.0         196                  1   
4                  15                    28.0         196                  1   
..                ...                     ...         ...                ...   
95                 12                     NaN         196                  1   
96                 17                     3.0         196                  9   
97                 10                    22.0         196                  3   
98                 19                    30.0         196                  2   
99                 19                    16.0         196                  2   

    reordered  ...    price_range_loc     busiest day    Busiest days  \
0           0  ...  Mid-range product  Regularly busy  Regularly busy   
1           1  ...  Mid-range product  Regularly busy      Least busy   
2           1  ...  Mid-range product  Regularly busy      Least busy   
3           1  ...  Mid-range product      Least busy      Least busy   
4           1  ...  Mid-range product      Least busy      Least busy   
..        ...  ...                ...             ...             ...   
95          0  ...  Mid-range product  Regularly busy  Regularly busy   
96          0  ...  Mid-range product  Regularly busy  Regularly busy   
97          0  ...  Mid-range product      Least busy      Least busy   
98          0  ...  Mid-range product  Regularly busy     Busiest day   
99          1  ...  Mid-range product  Regularly busy      Least busy   

    Busiest_period_of_day max_order      loyalty_flag mean_prices  \
0             Most orders        10      New customer    6.367797   
1           Fewest orders        10      New customer    6.367797   
2             Most orders        10      New customer    6.367797   
3           Fewest orders        10      New customer    6.367797   
4             Most orders        10      New customer    6.367797   
..                    ...       ...               ...         ...   
95            Most orders         3      New customer   10.006250   
96            Most orders         3      New customer    8.496552   
97            Most orders         8      New customer    7.396610   
98         Average orders        22  Regular customer    7.387805   
99         Average orders        22  Regular customer    7.387805   

    Spender_flag order_frequency         frequency_flag  
0    Low Spender            20.5  Non frequent customer  
1    Low Spender            20.5  Non frequent customer  
2    Low Spender            20.5  Non frequent customer  
3    Low Spender            20.5  Non frequent customer  
4    Low Spender            20.5  Non frequent customer  
..           ...             ...                    ...  
95  High Spender             4.0      Frequent customer  
96   Low Spender    

In [48]:
# check shape
op_merged.shape

(32404859, 25)

In [49]:
# check value counts
op_merged['frequency_flag'].value_counts(dropna = False)

Frequent customer        21559853
Regular customer          7208564
Non frequent customer     3636437
NaN                             5
Name: frequency_flag, dtype: int64

In [50]:
# we can see that 5 rows have missing data

In [52]:
# Export data to pkl
op_merged.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged_48task.pkl'))
